text/3396-extern-types-v2.md

```diff
@@ -0,0 +1,237 @@
+* dynamically - the size/alignment can only be determined at run time.
+  There are no types currently expressible in the language with dynamically known size or alignment.  
+  The most discussed potential type in this category is `CStr`, which has a statically known alignment but it's size can only be determined by iterating over it's contents to find the position of the null byte.
+  Note that these types are odd, for example determining the size of a `Mutex<CStr>` requires taking a lock on the mutex.
+* unknown - the size/alignment is not able to be determined at compile time or run time.  
+  This is the category that opaque types fall in (and no other existing types occupy), without any additional domain specific knowledge.
+  Therefore extern types will occupy this category to allow the most flexibility.
+
+The rest of this document will refer to types as "statically sized", "metadata aligned", etc.
+
+"dynamically aligned" (or "unknown aligned") types cannot be placed as the last field of a struct as their offset cannot be determined, without already having a pointer to the field.
+
+In the Rust 2021 edition and earlier `T: Sized` implies that `T` is "statically sized" and "statically aligned" and `T: ?Sized` implies that `T` is "metadata sized" and "metadata aligned".
+
+In the Rust 2024 edition and later `T: Sized` means the same but `T: ?Sized` implies that `T` is "unknown sized" and "unknown aligned".
+A `MetaSized` bound can be introduced to regain the previous meaning.
+
+## `MetaSized` trait
+[metasized-trait]: #metasized-trait
+
+This introduces a new trait `core::marker::MetaSized` that represents a type that is "metadata sized" and "metadata aligned":
+```rust
+#[lang = "meta_sized"]
+trait MetaSized: Pointee {
+    fn size_of_val(metadata: <Self as Pointee>::Metadata) -> usize;
+    fn align_of_val(metadata: <Self as Pointee>::Metadata) -> Alignment;
+}
+```
+This trait is automatically implemented for all types except extern types.
+
+All types that implement `MetaSized` can be placed as the last field in a struct because the compiler can emit code to determine the offset of the field purely from the pointer metadata.
+At the time of writing all `MetaSized` types are either statically aligned, or are trait objects and have their alignment in their vtable.
+This RFC does not propose changing codegen for computing the offset as it does not introduce any new `MetaSized` types.
+
+All locations in the standard library that use `?Sized` bounds need to be reviewed when migrating to the 2024 edition and replaced with either `?Sized` or `?Sized + MetaSized` as appropriate. Some examples:
+```rust
+pub fn size_of_val<T: ?Sized + MetaSized>(val: &T) -> usize
+```
+This requires `MetaSized` because the size must be known at run time.
+
+```rust
+pub struct Box<T: ?Sized + MetaSized, A: Allocator = Global>(Unique<T>, A);
```

Comment:
Without implied bounds, this will require the bound be mentioned everywhere `Box` (or `Arc` etc) are mentioned and can take a non-`Sized` parameter, including in traits.
```rust
pub trait Trait {
    fn foo(self: Box<Self>) where Self: MetaSized;
    fn bar(self: Arc<Self>) where Self: MetaSized;
}
```

Reply:
Interesting but probably bad idea: `MetaSized` as a default bound for all `trait`, unless they opt out.

It's worth noting that since receivers are special, we *could* add special rules to make a `Box<Self>` receiver imply `Self: MetaSized` in traits. On the other hand, we'd probably prefer to make receivers less special.

Reply:
[Implied bounds](https://github.com/rust-lang/rust/issues/44491) could solve this if they are ever implemented…

Reply:
This is really sad. It feels like a blocker that can't be easily skipped. How reasonable would it be to make receivers special until implied bounds is implemented? Or is the only option here to go and implement implied bounds?

Reply:
I just had a thought, doesn't implied bounds for self types arguably already exist given that you can write this:
```rust
trait Trait {
    fn foo(self);
}
```
You don't have to specify `where Self: Sized` on that. Granted, you can't implement it for unsized types but that feels like the correct behaviour.

Reply:
well, technically:
```rust
trait Trait {
    fn foo(self);
}
```
doesn't actually imply `where Self: Sized`, just that you need unstable features to implement it without `where Self: Sized`. for a good example, see `FnOnce` which can be used as `Box<dyn FnOnce(...)>` to call `fn call_once(self, ...)` with `Self = dyn FnOnce(...)`